File: leetcode/1-100/1-two-sum/bench/hash_map.py
"""Benchmark workload — hash-map O(n) Two Sum.

Same N / K / target as bench/brute_force.py. The single-pass dict
implementation walks the array once per call (O(n) work) — at
N=5000, K=10 that is ~50,000 ops per process. See ../README.md
§ Benchmarks for what the numbers mean.
"""

from __future__ import annotations


def two_sum(nums: list[int], target: int) -> tuple[int, int] | None:
    seen: dict[int, int] = {}
    for i, num in enumerate(nums):
        complement = target - num
        if complement in seen:
            return seen[complement], i
        seen[num] = i
    return None


def main() -> None:
    N = 5000
    data = [(i * 7) % 1000 for i in range(N)]
    target = -1

    sum_result = 0
    for _ in range(10):
        r = two_sum(data, target)
        if r is None:
            sum_result += -1 + -1
        else:
            sum_result += r[0] + r[1]
    print(sum_result)


if __name__ == "__main__":
    main()

File: leetcode/1-100/1-two-sum/bench/hash_map.rs
//! Benchmark workload — hash-map O(n) Two Sum.
//!
//! Algorithmic mirror of bench/hash_map.py. Uses std::collections::HashMap.
//! See ../README.md § Benchmarks for what the numbers mean.

use std::collections::HashMap;

fn two_sum(nums: &[i64], target: i64) -> Option<(usize, usize)> {
    let mut seen: HashMap<i64, usize> = HashMap::new();
    for (i, &num) in nums.iter().enumerate() {
        let complement = target - num;
        if let Some(&j) = seen.get(&complement) {
            return Some((j, i));
        }
        seen.insert(num, i);
    }
    None
}

fn main() {
    const N: usize = 5000;
    let data: [i64; N] = std::array::from_fn(|i| ((i as i64) * 7) % 1000);
    let target: i64 = -1;

    let mut sum: i64 = 0;
    for _ in 0..10 {
        sum += match two_sum(&data, target) {
            Some((i, j)) => i as i64 + j as i64,
            None => -2,
        };
    }
    println!("{}", sum);
}
